### PR TITLE
Clean-up Python #4815 - P2.1: Chem Module: A - C + DSViewer

### PR DESCRIPTION
**Pull request by IchiruTake**
This is the sub-PR breaking from the large PR 4815: https://github.com/rdkit/rdkit/pull/4815

--------------------------------
Updated Changes:
- [Remark]: In `ChemicalFeatures.py`, https://stackoverflow.com/questions/16940293/why-is-there-no-tuple-comprehension-in-python
- [Recall]: Optimize function `BRICSBuild()`
- [Warning]: An extra test has been added into `Mol3D.py` to test shortcut import in `AllChem.py`
- [Warning]: ChemSmarts.py - Line 32
- [New]: `time.perf_counter()` is used in `BuildFragmentCatalog.py` to replace `time.time()`
- [State]: I remove the variable assignment in line 297 in function `Zoom()` due to the similar syntax at line 266 and as well as there is no use in the function
- [New]: Function `SelectProteinNeighborhood()` can be slightly faster

### CODE REVIEW

File: rdkit/Chem/UnitTestCatalog.py
Comment (gedeck) [inline rdkit/Chem/UnitTestCatalog.py:160]: No need for `str(bit)`

File: rdkit/Chem/UnitTestCrippen.py
Comment (gedeck) [inline rdkit/Chem/UnitTestCrippen.py:87]:
```
f'bad logp fooutF,r {smi}: {clog:4.4f} != {tmp:4.4f}'
```

Comment (gedeck) [reply rdkit/Chem/UnitTestCrippen.py:87]: You're probably right with your change.

@greglandrum Why call `Crippen.MolLogP` twice?

File: rdkit/Chem/UnitTestMol3D.py
Comment (gedeck) [inline rdkit/Chem/UnitTestMol3D.py:459]: Where is this test coming from?

File: rdkit/Chem/__init__.py
Comment (gedeck) [inline rdkit/Chem/__init__.py:45]: I leave it to @greglandrum to comment on adding type hints.

File: rdkit/Chem/__init__.py
Comment (gedeck) [inline rdkit/Chem/__init__.py:137]: !!!!!  This bypasses everything below it. I don't think this is what should happen.

**Review by gedeck — CHANGES_REQUESTED** [review]

File: rdkit/Chem/AllChem.py
Comment (IchiruTake) [reply rdkit/Chem/AllChem.py:343]: I think the for-loop is more simple than while loop. Yes, it runed 5 times as in the original code which runs 1 in the outer `while`. Then it runs at n=4, n=3, n=2, and n=1. n=0 is ignored.
Thank you for your review.

File: rdkit/Chem/BRICS.py
Comment (IchiruTake) [reply rdkit/Chem/BRICS.py:218]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/BRICS.py
Comment (IchiruTake) [reply rdkit/Chem/BRICS.py:305]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/BRICS.py
Comment (IchiruTake) [reply rdkit/Chem/BRICS.py:484]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/BRICS.py
Comment (IchiruTake) [reply rdkit/Chem/BRICS.py:521]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/BuildFragmentCatalog.py
Comment (IchiruTake) [reply rdkit/Chem/BuildFragmentCatalog.py:287]: Just to add thing a little bit clear that we don't do this code in the first iteration. But if you want, I will revert it.

File: rdkit/Chem/BuildFragmentCatalog.py
Comment (IchiruTake) [reply rdkit/Chem/BuildFragmentCatalog.py:179]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/BuildFragmentCatalog.py
Comment (IchiruTake) [reply rdkit/Chem/BuildFragmentCatalog.py:363]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/ChemicalFeatures.py
Comment (IchiruTake) [reply rdkit/Chem/ChemicalFeatures.py:16]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/UnitTestCatalog.py
Comment (IchiruTake) [reply rdkit/Chem/UnitTestCatalog.py:160]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/UnitTestCrippen.py
Comment (IchiruTake) [reply rdkit/Chem/UnitTestCrippen.py:87]: > ```
> f'bad logp fooutF,r {smi}: {clog:4.4f} != {tmp:4.4f}'
> ```

I don't think it is right. While other code call `bad logp` right but at this time not? Related to the second answer, the test named `testRepeat` when he want to guarantee the code does not change it behaviour if called twice, I guessed.

File: rdkit/Chem/UnitTestMol3D.py
Comment (IchiruTake) [reply rdkit/Chem/UnitTestMol3D.py:459]: As mentioned earlier, this came from the `import`. This test is important as we want these two functions can be accessed shortcut on `AllChem.py` but not only `Chem.EnumerateStereoisomers.py`. This is to prevent people try to delete the shortcut import on `AllChem.py`. If it is redundant, I can delete it.

Comment (IchiruTake) [reply rdkit/Chem/UnitTestMol3D.py:459]: If you dig out `AllChem.py`, you would see that a numerous of function has been imported but not use in the module.

File: rdkit/Chem/__init__.py
Comment (IchiruTake) [reply rdkit/Chem/__init__.py:137]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/UnitTestMol3D.py
Comment (gedeck) [reply rdkit/Chem/UnitTestMol3D.py:455]: I get it. I wonder if there is a simpler test. One could for example do the following:
```
from rdkit import Chem
from rdkit.Chem import AllChem
chem = set(Chem.__dict__)
allchem = set(AllChem.__dict__)

expectedMissing = {'AllChem'}
chem = chem - expectedMissing
missing = [k for k in chem if k not in allchem and not k.startswith('_')]
... filter list of functions contained in chem but not available in allchem ...
missing = 
assert not missing
```
Currently this returns:
```
>>> missing
['rdMolAlign', 'rdDepictor', 'rdMolTransforms', 'rdReducedGraphs', 'rdPartialCharges', 'rdChemReactions', 'rdSLNParse', 'rdShapeHelpers', 'rdMolEnumerator', 'rdForceFieldHelpers', 'ChemicalFeatures', 'rdMolChemicalFeatures', 'rdDistGeom', 'rdMolDescriptors', 'rdChemicalFeatures', 'rdqueries']
```
@greglandrum looks like not all of the functionality is exported by `AllChem`. Would it make sense to add a test like this to make sure that new functionality is also added to `AllChem`?

File: rdkit/Chem/UnitTestCrippen.py
Comment (gedeck) [reply rdkit/Chem/UnitTestCrippen.py:87]: Right. missed that. In this case, I would add an assert after each call.

**Review by gedeck — APPROVED** [review]
@greglandrum a few comments for you to consider.

File: rdkit/Chem/UnitTestMol3D.py
Comment (IchiruTake) [reply rdkit/Chem/UnitTestMol3D.py:455]: > here is a simpler test that could be added to `UnitTestAllChem`. One could for example do the following:

I don't recommend this test, as a slight change in `AllChem` but not in `Chem` would break the test and everyone has to add it in the module. Meanwhile, although it has stated that `AllChem` is the extension of `Chem` but aimed at the more complex operation. I doubt that when we add functions into `AllChem`, this could be break

Comment (gedeck) [reply rdkit/Chem/UnitTestMol3D.py:455]: If you want to make sure that functionality is exposed through `AllChem` and `Chem` then having a test in `UnitTestAllChem` that does this is a reasonable approach. Adding functionality to `AllChem` would not break the suggested test, but adding functionality to `Chem` would. At this point, one would need to address this situation either by including the export into `AllChem` or by adding the new functionality to the `expectedMissing` list. In this case, breaking the test is what you want to have.

Comment (IchiruTake) [reply rdkit/Chem/UnitTestMol3D.py:455]: I think the better case is to delete the test I have made. This would be better.

Comment (gedeck) [reply rdkit/Chem/UnitTestMol3D.py:455]: I agree. I'll point Greg to this discussion to see if he thinks adding the AllChem test (and potentially adding the missing imports) would be useful.

File: rdkit/Chem/__init__.py
Comment (greglandrum) [reply rdkit/Chem/__init__.py:45]: I am 100% opposed to type hints

Comment (greglandrum) [reply rdkit/Chem/__init__.py:45]: I'm not going to do that particular rant about why here, but code which uses type hints is inconsistent with the rest of the code base and I won't merge it.

File: rdkit/Chem/UnitTestMol3D.py
Comment (greglandrum) [reply rdkit/Chem/UnitTestMol3D.py:455]: The original intent of AllChem is that it should bring in all the extension modules. And, you're right that it no longer does so.
I think fixing this and adding the test to automatically catch future problems is a great idea.

Comment (IchiruTake) [reply rdkit/Chem/UnitTestMol3D.py:455]: > The original intent of AllChem is that it should bring in all the extension modules. And, you're right that it no longer does so. I think fixing this and adding the test to automatically catch future problems is a great idea.

I agree, but I don't have enough time to check every extension is correctly imported into `AllChem.py`. Eventhough it could, I have seen tons of `AllChem.MolFromSmiles()` and similar function which should not be called but should on `Chem.py`.

File: rdkit/Chem/__init__.py
Comment (IchiruTake) [reply rdkit/Chem/__init__.py:45]: OK, I got your idea! I will revert it back to original

File: rdkit/Chem/UnitTestMol3D.py
Comment (greglandrum) [reply rdkit/Chem/UnitTestMol3D.py]: > I agree, but I don't have enough time to check every extension is correctly imported into `AllChem.py`.

@gedeck showed an example above of how to do this.

Comment (IchiruTake) [reply rdkit/Chem/UnitTestMol3D.py]: > > I agree, but I don't have enough time to check every extension is correctly imported into `AllChem.py`.
> 
> @gedeck showed an example above of how to do this.

Thanks @gedeck, but I believe this beyond the scope of this PR, I will make a second one to tackle with this. I had also revert the type hint in `__init__.py` back to original. I also switch the `LegacyImplementation` of `FindMolChiralCenters to different function. This can help another developers understand and did not make any unintentional violation

**Comment by IchiruTake** [comment]
@greglandrum I have resolved the conflict in PR 5014

**Pull request closed without merging**
